### PR TITLE
Possible fix for issue 481

### PR DESCRIPTION
**Pull request by naneer**
I originally thought it was due to the display property being set to none for the panels, which is sort of true, but I believe now that the root cause is the panel's animation event listener.  It was picking up the transitionend events for the child zf-open button element, it transform's the background-color property.  

Since the transition for the button color occurs rather quickly, the panel runs finishAnimation too soon, setting the display property to none and canceling the transitionend event for the the child panel.

All in all, I think it be tighter if the element would verify that the `transitionend` event to execute on is for the transition on itself as the target.  Allowing panels to support child elements with transitions/animations without effecting it's own.

Instead of element.one(), use .on() and .off() to allow the eventHandler to find the correct transition to run on.  Also, I set a clearTimeout in the handler to stop the setTimeout function from running if it doesn't need to.  Thinking about using $timeout if necessary in the future.  I also changed it to just check if same matching id.

````javascript
      element.on(events.join(' '), eventHandler);

      function eventHandler(e) {
        if (element[0].id === e.target.id) {
          clearTimeout(animationTimeout);
          finishAnimation();
        }
      }

      var animationTimeout = setTimeout(function() {
        if(timedOut) {
          finishAnimation();
        }
      }, 3000);

      function finishAnimation() {
        deregisterElement(element);
        reset(); //reset all classes
        element[0].style.transitionDuration = '';
        element.removeClass(!activation ? activeGenericClass : ''); //if not active, remove active class
        reflow();
        timedOut = false;
        element.off(events.join(' '), eventHandler);
      }
````

### CODE REVIEW

**Comment by gjungb** [comment]
+1 for this

Are there any plans to merge this pull request? Unfortunately the ```zf-panel``` is too buggy to be used even with a single ```.button``` inside of it.

File: js/angular/services/foundation.core.animation.js [hunk 1]
```diff
@@ -56,16 +56,21 @@
...
 
-      setTimeout(function() {
+      var animationTimeout = setTimeout(function() {
         if(timedOut) {
           finishAnimation();
         }
       }, 3000);
 
+      function eventHandler(e) {
+        if (element[0].id === e.target.id) {
```
Comment (gjungb) [inline js/angular/services/foundation.core.animation.js]: Wouldn't it be more verbose not to rely on the ```id``` but to use the comparison ```this === e.target``` as proposed in the jQuery docs (http://api.jquery.com/event.target/)?
> It is often useful to compare event.target to this in order to determine if the event is being handled due to event bubbling.

Comment (naneer) [inline js/angular/services/foundation.core.animation.js]: Hey thanks for the +1, `this` may equal to the animation service, would reverting back to `element[0] === e.target` be enough to follow the jQuery doc?

Comment (gjungb) [inline js/angular/services/foundation.core.animation.js]: @naneer yes, makes sense to me

**Comment by ViniciusTavares** [comment]
I update my foundation-apps source but the problem continues with modals. 
When I have a modal with a button then it is closing automatically.

Panels are ok.

**Comment by heppy83** [comment]
I'm still experiencing this problem both with panels and modals.
More precisely, if I have a panel (or modal) with a button inside, when I:
1. open the panel (with zf-open directive or using the FoudationApi methods )
2. close the panel - with zf-close directive or with a custom close link (using the FoudationApi methods)
3. reopen the panel 
4. hover the button inside the panel

At this point the (i.e., as soon as I hover the button), the panel disappears (is closed with no animation), and cannot be reopened even triggering the open button used to open it the first time.

I see this issue is closed and issue #598 references to this issue, but still I have this problems.

Is there a a workaround for this? Or perhaps are you currently working on it and it is expected to be fixed in next releases?

IMPORTANT NOTE: I DO see the problems on Chrome ( 45.0.2454.101) and NOT on Firefox (41.0.2).

**Comment by DAmatheson** [comment]
I've just run into what I think is this issue myself. 

This root cause of the issue seen [here at lines 158-159](https://github.com/zurb/foundation-apps/blob/master/js/angular/components/panel/panel.js#L158-L159) in `panel.js`.

In my case, it is caused by `FoundationAnimation.animate`'s `run()` being called twice. This means the `animationTimeout` for the first run isn't cancelled by `eventHandler()` when it finishes. 

The first call is coming from `scope.hide` in the `postlink` of `zfPanel`. The second call is coming from `element.on('click' ...)` which is also the original caller to `scope.hide`. As both `scope.hide` and `element.on` call animate, you get two animations running and the second one overrides the `animationTimeout` token for the first one which means the first one can't be cancelled.

Short version: zfPanel's on click calls both `scope.hide()` and `animate`. This is bad as the animations overlap and the first one's timeout fails to cancel.

**Comment by soumak77** [comment]
@DAmatheson nice catch!  I've submitted a PR for the fix.
